frame/evm/src/runner/stack.rs

```diff
@@ -130,18 +130,23 @@ where
 			>,
 		) -> (ExitReason, R),
 	{
-		// EIP-3607: https://eips.ethereum.org/EIPS/eip-3607
-		// Do not allow transactions for which `tx.sender` has any code deployed.
-		//
-		// We extend the principle of this EIP to also prevent `tx.sender` to be the address
-		// of a precompile. While mainnet Ethereum currently only has stateless precompiles,
-		// projects using Frontier can have stateful precompiles that can manage funds or
-		// which calls other contracts that expects this precompile address to be trustworthy.
-		if !<AccountCodes<T>>::get(source).is_empty() || precompiles.is_precompile(source) {
-			return Err(RunnerError {
-				error: Error::<T>::TransactionMustComeFromEOA,
-				weight,
-			});
+		// Only check the restrictions of EIP-3607 if the source of the EVM operation is from an external transaction.
+		// If the source of this EVM operation is from an internal call, like from `eth_call` or `eth_estimateGas` RPC,
+		// we will skip the checks for the EIP-3607.
+		if is_transactional {
+			// EIP-3607: https://eips.ethereum.org/EIPS/eip-3607
+			// Do not allow transactions for which `tx.sender` has any code deployed.
+			//
+			// We extend the principle of this EIP to also prevent `tx.sender` to be the address
+			// of a precompile. While mainnet Ethereum currently only has stateless precompiles,
+			// projects using Frontier can have stateful precompiles that can manage funds or
+			// which calls other contracts that expects this precompile address to be trustworthy.
+			if !<AccountCodes<T>>::get(source).is_empty() || precompiles.is_precompile(source) {
+				return Err(RunnerError {
+					error: Error::<T>::TransactionMustComeFromEOA,
+					weight,
+				});
+			}
 		}
```

Comment:
```suggestion
		    // Only check the restrictions of EIP-3607 if the source of the EVM operation is from an external transaction.
		    // If the source of this EVM operation is from an internal call, like from `eth_call` or `eth_estimateGas` RPC,
		    // we will skip the checks for the EIP-3607.
		    //
			// EIP-3607: https://eips.ethereum.org/EIPS/eip-3607
			// Do not allow transactions for which `tx.sender` has any code deployed.
			//
			// We extend the principle of this EIP to also prevent `tx.sender` to be the address
			// of a precompile. While mainnet Ethereum currently only has stateless precompiles,
			// projects using Frontier can have stateful precompiles that can manage funds or
			// which calls other contracts that expects this precompile address to be trustworthy.
			if is_transactional && !<AccountCodes<T>>::get(source).is_empty() || precompiles.is_precompile(source) {
				return Err(RunnerError {
					error: Error::<T>::TransactionMustComeFromEOA,
					weight,
				});
			}
```

Reply:
It seems that the moonbeam developers have extended the principle of EIP-3607, should I change it to something like this?

Reply:
Good question, I'd say in non-transactional context we don't care, @nanocryk to confirm

Reply:
I did some tests with Sepolia and `eth_call` allowed  the `from` to be a contract. But the same `eth_call` 
 reverted on Moonbeam with `TransactionMustComeFromEOA`

Reply:
Sounds good to me

Reply:
> Sounds good to me

@nanocryk so calling non-transactional call with stateful precompiles as the source will be fine, right?